src/pre-charge/pre_charge.cpp

```diff
@@ -0,0 +1,257 @@
+#include <pre-charge/pre_charge.hpp>
+
+#include <EVT/utils/time.hpp>
+
+namespace IO = EVT::core::IO;
+namespace time = EVT::core::time;
+
+namespace pre_charge {
+
+pre_charge::pre_charge(IO::GPIO& key, IO::GPIO& batteryOne, IO::GPIO& batteryTwo,
+                       IO::GPIO& eStop, IO::GPIO& pc, IO::GPIO& dc, IO::GPIO& cont,
+                       IO::GPIO& apm, IO::GPIO& forward) : key(key),
+                                                           batteryOne(batteryOne),
+                                                           batteryTwo(batteryTwo),
+                                                           eStop(eStop),
+                                                           pc(pc),
+                                                           dc(dc),
+                                                           cont(cont),
+                                                           apm(apm),
+                                                           forward(forward) {
+    state = State::MC_OFF;
+
+    keyStatus = IO::GPIO::State::LOW;
+    stoStatus = IO::GPIO::State::LOW;
+    batteryOneStatus = IO::GPIO::State::LOW;
+    batteryTwoStatus = IO::GPIO::State::LOW;
+    eStopStatus = IO::GPIO::State::LOW;
+    //TODO: Add GFD
+}
+
+void pre_charge::handle() {
+    getSTO();  //update value of STO
+    getMCKey();//update value of MC_KEY_IN
+
+    switch (state) {
+    case pre_charge::State::MC_OFF:
+        mcOffState();
+        break;
+    case pre_charge::State::ESTOPWAIT:
+        eStopState();
+        break;
+    case pre_charge::State::PRECHARGE:
+        prechargeState();
+        break;
+    case pre_charge::State::CONT_CLOSE:
+        contCloseState();
+        break;
+    case pre_charge::State::MC_ON:
+        mcOnState();
+        break;
+    case pre_charge::State::CONT_OPEN:
+        contOpenState();
+        break;
+    case pre_charge::State::DISCHARGE:
+        dischargeState();
+        break;
+    case pre_charge::State::FORWARD_DISABLE:
+        forwardDisableState();
+    default:
+        break;
+    }
+}
+
+void pre_charge::getSTO() {
+    batteryOneStatus = batteryOne.readPin();
+    batteryTwoStatus = batteryTwo.readPin();
+    eStopStatus = eStop.readPin();
+    //TODO: Add GFD
+
+    if (batteryOneStatus == IO::GPIO::State::HIGH && batteryTwoStatus == IO::GPIO::State::HIGH && eStopStatus == IO::GPIO::State::LOW) {
+        stoStatus = IO::GPIO::State::HIGH;
+    } else {
+        stoStatus = IO::GPIO::State::LOW;
+    }
+}
+
+void pre_charge::getMCKey() {
+    keyStatus = key.readPin();
+}
+
+void pre_charge::setPrecharge(int state) {
+    if (state) {
+        pc.writePin(IO::GPIO::State::HIGH);
+    } else {
+        pc.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::setDischarge(int state) {
+    if (state) {
+        dc.writePin(IO::GPIO::State::HIGH);
```

Comment:
If you change the `state` variable to a `uint8_t` you can then static cast the `state` to a GPIO state